### PR TITLE
primary key strings!

### PR DESCRIPTION
**Pull request by osheroff**
the fruition of all of last week's labor.

@vanchi-zendesk

### CODE REVIEW

File: src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java [hunk 1]
```diff
@@ -242,4 +243,33 @@ public Object getData(String string) {
...
+		for ( String pk : table.getPKList() ) {
+			int idx = table.findColumnIndex(pk);
+
+			Column column = r.getColumns().get(idx);
+			ColumnDef def = table.getColumnList().get(idx);
+
+
+			pkValueList.add(def.getName() + ":" + def.asJSON(column.getValue()).toString());
+		}
+		return StringUtils.join(pkValueList, "/");
```
Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java]: values can contain both ":" and "/", just saying.

Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java]: there's no separator they can't contain.  Only other option I could think of would be casting the whole key to json....

File: src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java [hunk 2]
```diff
@@ -242,4 +243,33 @@ public Object getData(String string) {
 		}
 		return list;
 	}
+
+	public List<String> getPKStrings() {
+		ArrayList<String> list = new ArrayList<>();
+
+		String pkString = "db:" + table.getDatabase().getName() + "/"
+				        + "tbl:" + table.getName() + "/";
```
Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java]: this is counter-intuitive. You expect getPKStrings to get you PK strings only, but sure.

**Comment by vanchi-zendesk** [comment]
:+1: :clap:

File: src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java (hunk 1 above)
Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java]: that's fine by me.. I mean, let JSON encoding worry about skipping stuff. Just create a map and do toJSON?

Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java]: Also, a json key is much easier to parse out.

Comment (osheroff) [inline src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java]: not quite as simple as just creating a map -- you gotta also ensure key ordering... but ok.

Comment (vanchi-zendesk) [inline src/main/java/com/zendesk/maxwell/MaxwellAbstractRowsEvent.java]: Oh, a list of [ { col: val }, { col: val } ] I guess...

File: docs/docs/kafka.md [hunk 3]
```diff
@@ -11,6 +11,18 @@ Maxwell sets the following Kafka options by default, but you can override them i
...
+
+Maxwell generates keys for its Kafka messages based upon a mysql row's primary key:
+
+```
+db:test_db/tbl:test_tbl/id:93/id_2:17910090
+```
+
+This key is designed to co-operate with Kafka's log compaction, which will save the last-known
+value for a key, allowing Maxwell's Kafka stream to retain the last-known value for a row and act
+as a source of truth.
```
Comment (vanchi-zendesk) [inline docs/docs/kafka.md]: Might want to explain this for tables without primary keys.

**Comment by osheroff** [comment]
@vanchi-zendesk  now with JSON keys.

**Comment by vanchi-zendesk** [comment]
:+1:

**Pull request merged**